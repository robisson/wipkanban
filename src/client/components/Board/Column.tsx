import * as React from "react";
import { withStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import Paper from "@material-ui/core/Paper";
import IconButton from "@material-ui/core/IconButton";
import Card from "./Card";
import ColumnHeader from "./ColumnHeader";
import ColumnBoard from "./Column";
import AddIcon from "@material-ui/icons/AddCircle";

type Props = {
  classes?: any;
  title: string;
  cards?: any;
  vColumns?: any;
  hColumns?: any;
};

type PropsContainer = {
  children: React.Component<typeof Column>;
};

function ColumnContainer(props: PropsContainer) {
  let { children } = props;

  return (
    <Grid container justify="flex-start" spacing={0}>
      {children}
    </Grid>
  );
}

function Column(props: Props) {
  const { title, classes, cards, vColumns, hColumns } = props;

  let children = null;

  if (vColumns && vColumns.length) {
    children = (
      <ColumnContainer>
        {vColumns.map((column: any) => (
          <ColumnBoard
            key={column.id}
            vColumns={column.vColumns}
            hColumns={column.hColumns}
            title={column.title}
            cards={column.cards}
          />
        ))}
      </ColumnContainer>
    );
  } else if (hColumns.length) {
    children = hColumns.map((column: any) => (
      <ColumnBoard
        key={column.id}
        vColumns={column.vColumns}
        hColumns={column.hColumns}
        title={column.title}
        cards={column.cards}
      />
    ));
  } else {
    children = cards.map((card: any) => (
      <Card key={card.id} title={card.title} />
    ));
  }

  return (
    <Grid item className={classes.column}>
      <Paper className={classes.parentColumn} elevation={0}>
        <ColumnHeader title={title} /> {children}
        {!hColumns.length && !vColumns.length ? (
          <div style={{ textAlign: "center" }}>
            <IconButton color="default" aria-label="Add card">
              <AddIcon />
            </IconButton>
          </div>
        ) : (
          ""
        )}
      </Paper>
    </Grid>
  );
}

function getWidthColumn(columns: any, defaultWidth: number = 270): any {
  return columns.map((column: any) => {
    return column.vColumns.length
      ? getWidthColumn(column.vColumns).reduce(
          (acum: number, current: number) => current + acum
        ) + column.vColumns.length
      : defaultWidth;
  });
}

function ColumnFactory(props: Props) {
  const { vColumns, hColumns } = props;

  let vWidth: number = 0;
  if (vColumns.length) {
    vWidth =
      getWidthColumn(vColumns).reduce(
        (acum: number, current: number) => current + acum
      ) + vColumns.length;
  } else if (hColumns.length) {
    vWidth = Math.max.apply(null, getWidthColumn(hColumns));
  } else {
    vWidth = 270;
  }

  const styles: any = () => ({
    parentColumn: {
      height: hColumns.length ? 450 : "100%"
    },
    column: {
      width: vWidth,
      height: hColumns.length ? 450 : "100%",
      borderStyle: "solid",
      borderWidth: 1,
      borderColor: "#e0e0e0"
    }
  });

  let Component = withStyles(styles)(Column);

  return <Component {...props} />;
}

export default ColumnFactory;
