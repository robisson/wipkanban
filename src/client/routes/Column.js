// @flow
import * as React from "react";
import {withStyles} from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import Card from './card';
import ColumnHeader from "./ColumnHeader";
import ColumnBoard from "./Column";

type Props = {
  classes: Object,
  title: string,
  cards?: Array < Card >,
  vColumns?: Array < Column >,
  hColumns?: Array < Column >
}

type PropsContainer = {
  children: React.Element < Column >
}

function ColumnContainer(props : PropsContainer) {

  let {children} = props;

  return (
    <Grid container justify="flex-start" spacing={0}>
      {children}
    </Grid>
  )
}

function Column(props : Props) {
  const {title, classes, cards, vColumns, hColumns} = props;

  let children = null;

  if (vColumns.length) {
    children = (
      <ColumnContainer>
        {vColumns.map(column => (<ColumnBoard
          key={column.id}
          vColumns={column.vColumns}
          hColumns={column.hColumns}
          title={column.title}
          cards={column.cards}/>))}
      </ColumnContainer>
    );
  } else if (hColumns.length) {
    children = 'hCOlumns';
  } else {
    children = cards.map(card => (<Card key={card.id} title={card.title}/>));
  }

  return (
    <Grid item className={classes.column}>
      <Paper className={classes.parentColumn} elevation={0}>
        <ColumnHeader title={title}/> {children}
      </Paper>
    </Grid>
  );
}

function getWidthColumn(columns, defaultWidth = 270) {

  return columns.map(column => {
    return column.vColumns.length
      ? getWidthColumn(column.vColumns).reduce((acum, current) => current + acum) + column.vColumns.length
      : defaultWidth;

  });

}

function ColumnFactory(props : Props) {

  const {vColumns} = props;

  /*if(vColumns.length){
    let widthTeste = getWidthColumn(vColumns).reduce((acum, current) => current + acum);
    widthTeste +=vColumns.length;
    console.log(`Largura da coluna ${props.title}`,widthTeste);
  }*/

  let vWidth = (vColumns.length
    ? getWidthColumn(vColumns).reduce((acum, current) => current + acum)+ vColumns.length
    : 270);

  const styles : Object = {
    parentColumn: {
      paddingBottom: 5
    },
    column: {
      width: vWidth,
      height: 1000,
      borderStyle: 'solid',
      borderWidth: 1,
      borderColor: '#e0e0e0'
    }
  };

  let Component = withStyles(styles)(Column)

  return <Component {...props}/>
}

export default ColumnFactory;
