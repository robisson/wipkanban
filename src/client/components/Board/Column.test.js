import React from "react";
import Column from "./Column";
import { createShallow } from "@material-ui/core/test-utils";
import toJson from "enzyme-to-json";
import columns from "./fakeData";

// setup file
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("<Column /> Component", () => {
  let shallow;
  let wrapper;

  beforeEach(() => {
    shallow = createShallow({ dive: true });
  });

  it("should render correctly", () => {
    wrapper = shallow(<Column {...columns[0]} />);

    expect(wrapper.name()).toEqual("Column");
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
