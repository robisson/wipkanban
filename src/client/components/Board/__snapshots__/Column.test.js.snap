// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Column /> Component should render correctly 1`] = `
<Column
  cards={
    Array [
      Object {
        "id": 31,
        "title": "Criar layouts da tela principal do quadro. Este Ã© um exemplo de tarefa ou card do quadro kanban",
      },
    ]
  }
  classes={
    Object {
      "column": "Column-column-2",
      "parentColumn": "Column-parentColumn-1",
    }
  }
  hColumns={Array []}
  id={4}
  title="A fazer"
  vColumns={Array []}
/>
`;
